akka-cluster/src/main/scala/akka/cluster/MembershipState.scala

```diff
@@ -102,16 +102,17 @@ import scala.util.Random
   /**
    * @return Up to `crossDcConnections` oldest members for each DC
    */
-  lazy val ageSortedTopOldestMembersPerDc: Map[DataCenter, SortedSet[Member]] =
-    // TODO make this recursive and bail early when size reached to make it fast for large clusters
+  lazy val ageSortedTopOldestMembersPerDc: Map[DataCenter, SortedSet[Member]] = {
+    // FIXME optimize for large clusters with small numbers of cross DC connections
     latestGossip.members.foldLeft(Map.empty[DataCenter, SortedSet[Member]]) { (acc, member) ⇒
       acc.get(member.dataCenter) match {
         case Some(set) ⇒
-          if (set.size < crossDcConnections) acc + (member.dataCenter → (set + member))
-          else acc
-        case None ⇒ acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
+          acc + (member.dataCenter → (set + member))
+        case None ⇒
+          acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
       }
-    }
+    }.mapValues(_.take(crossDcConnections))
```

Comment:
`mapValues` creates a view, which we should perhaps stay away from?

Reply:
Yeah i'll change it to the version that assumes that if the cluster is large the number of cross DC conns is small

Reply:
Also forgot to checkin the test for this, doh